### ISSUE TITLE
MJWarp and PyTorch autograd support

### ISSUE THREAD

**Issue by zhyii-baa** [google-deepmind/mujoco_warp#1292]
Hi,

I would like to confirm whether MJWarp currently supports end-to-end gradient propagation with PyTorch.

My use case is the following:

- I have a neural network in PyTorch that predicts some control-related quantity.
- During training, I need to call MuJoCo/MJWarp inside the forward pass to advance the system state.
- Then I compute a loss from the simulated next state, and I want to backpropagate that loss all the way to the PyTorch network parameters.

For example, suppose I have:

tau = net(obs)                     # PyTorch tensor
next_state = mjwarp_step(state, tau)
loss = loss_fn(next_state, target)
loss.backward()

Should I expect gradients from `loss` to flow back to `tau` and then to `net.parameters()`?
Or is this currently unsupported in MJWarp?

My question is:

1. Even if I pass data from PyTorch to Warp arrays, does MJWarp support gradients through the simulation step itself?
2. Or is MJWarp currently only suitable for forward simulation, while the simulation step is not differentiable with respect to PyTorch parameters?
3. If it is not differentiable, is there any recommended way to integrate MJWarp into a training loop that requires gradients?
4. Are there any plans to support differentiable simulation / PyTorch autograd in the future?

The practical issue I am facing is that, in the loss computation, I need to call MuJoCo simulation, but if the simulation step is not part of a differentiable PyTorch computation graph, then gradients cannot flow back to the neural network.

Thanks.

**Comment by thowell** [comment]
autodifferentiation is not currently available for mujoco warp. this feature is tracked in issue #500.

**Comment by zhyii-baa** [comment]
Oh, I understand. Thanks for replying！
